### PR TITLE
LibJS: Spec compliant CreateDynamicFunction

### PR DESCRIPTION
**Pull request by linusg**
```
Diff Tests:
    test/built-ins/AsyncGeneratorFunction/instance-await-expr-in-param.js                             📝 -> ❌
    test/built-ins/Function/prototype/toString/GeneratorFunction.js                                   💥️ -> ❌
    test/built-ins/GeneratorFunction/has-instance.js                                                  💥️ -> ✅
    test/built-ins/GeneratorFunction/instance-construct-throws.js                                     💥️ -> ✅
    test/built-ins/GeneratorFunction/instance-length.js                                               💥️ -> ✅
    test/built-ins/GeneratorFunction/instance-name.js                                                 💥️ -> ✅
    test/built-ins/GeneratorFunction/instance-prototype.js                                            💥️ -> ❌
    test/built-ins/GeneratorFunction/instance-restricted-properties.js                                💥️ -> ✅
    test/built-ins/GeneratorFunction/instance-yield-expr-in-param.js                                  💥️ -> 📝
    test/built-ins/GeneratorFunction/invoked-as-constructor-no-arguments.js                           💥️ -> 📝
    test/built-ins/GeneratorFunction/invoked-as-function-multiple-arguments.js                        💥️ -> 📝
    test/built-ins/GeneratorFunction/invoked-as-function-no-arguments.js                              💥️ -> 📝
    test/built-ins/GeneratorFunction/invoked-as-function-single-argument.js                           💥️ -> 📝
    test/built-ins/GeneratorFunction/is-a-constructor.js                                              💥️ -> ✅
    test/built-ins/GeneratorFunction/proto-from-ctor-realm-prototype.js                               💥️ -> ❌
    test/built-ins/GeneratorFunction/proto-from-ctor-realm.js                                         💥️ -> ❌
    test/built-ins/Object/seal/seal-generatorfunction.js                                              💥️ -> ✅
    test/language/statements/class/subclass/builtin-objects/GeneratorFunction/instance-length.js      💥️ -> ✅
    test/language/statements/class/subclass/builtin-objects/GeneratorFunction/instance-name.js        💥️ -> ✅
    test/language/statements/class/subclass/builtin-objects/GeneratorFunction/instance-prototype.js   💥️ -> ✅
    test/language/statements/class/subclass/builtin-objects/GeneratorFunction/regular-subclassing.js  💥️ -> 📝
    test/language/statements/class/subclass/builtin-objects/GeneratorFunction/super-must-be-called.js 💥️ -> ✅
```

### CODE REVIEW

File: Userland/Libraries/LibJS/Parser.cpp [hunk 1]
```diff
@@ -2470,16 +2470,16 @@ NonnullRefPtr<FunctionNodeType> Parser::parse_function_node(u8 parse_options)
         function_kind = FunctionKind::Normal;
     String name;
     if (parse_options & FunctionNodeParseOptions::CheckForFunctionAndName) {
-        if (function_kind == FunctionKind::Normal && match(TokenType::Async) && !next_token().trivia_contains_line_terminator()) {
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Parser.cpp]: Is this change intended?

File: Userland/Libraries/LibJS/Parser.h [hunk 2]
```diff
@@ -297,5 +297,11 @@ class Parser {
     Vector<ParserState> m_saved_state;
     HashMap<Position, TokenMemoization, PositionKeyTraits> m_token_memoizations;
     Program::Type m_program_type;
+
+public:
+    // Hidden at the very end so noone will be tempted to use it. Seriously, please don't.
+    // This should only ever be needed by FunctionConstructor::create_dynamic_function_node() for some manual parser state shenanigans.
+    ParserState& state() { return m_state; };
```
Comment (davidot) [inline Userland/Libraries/LibJS/Parser.h:304]: It is uglier but if it is really just one function which uses it we could just make this a friend?
Because I agree this state being visible can be dangerous.

File: Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp [hunk 3]
```diff
@@ -37,32 +38,231 @@ FunctionConstructor::~FunctionConstructor()
...
+    }
+
+    // 21. Let body be ParseText(! StringToCodePoints(bodyString), bodySym).
+    bool contains_direct_call_to_eval = false;
+    auto body_parser = Parser { Lexer { body_string } };
+    // Set up some parser state to accept things like return and yield.
+    body_parser.state().in_function_context = true;
+    if (parse_options & FunctionNodeParseOptions::IsGeneratorFunction)
+        body_parser.state().in_generator_function_context = true;
+    (void)body_parser.parse_function_body(parameters, kind, contains_direct_call_to_eval);
```
Comment (davidot) [inline Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp:199]: Do we not need to set `await_expression_is_valid` for async functions here?

**Review by davidot — COMMENTED** [review]
LGTM just a small comment about the `state()` method on parser

File: Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp [hunk 4]
```diff
@@ -37,32 +38,231 @@ FunctionConstructor::~FunctionConstructor()
...
     }
 
+    // 27. Let proto be ? GetPrototypeFromConstructor(newTarget, fallbackProto).
+    auto* prototype = TRY(get_prototype_from_constructor(global_object, *new_target, fallback_prototype));
+
+    // 28. Let realmF be the current Realm Record.
+    auto* realm = vm.current_realm();
+
+    // 29. Let scope be realmF.[[GlobalEnv]].
+    Environment* scope = &realm->global_environment();
```
Comment (IdanHo) [inline Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp:228]: Any specific reason for explicitly naming the type?

File: Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp (hunk 3 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp:199]: Oh and yes that name is different because I eventually wanted to move toward mimicking the `[~Await] [+Yield]` like names instead of in_generator.

(Also since await can be valid outside async functions)

**Review by IdanHo — COMMENTED** [review]
Nice work, just a couple o' nitpicks

File: Userland/Libraries/LibJS/Parser.cpp (hunk 1 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Parser.cpp]: Yeah this might mess with the valid
`async function async() { await 1; }`
Which we don't have a test for... Whoops that's my bad

Comment (IdanHo) [reply Userland/Libraries/LibJS/Parser.cpp]: Sure, but it doesn't seem related to the commit which changes it.

Comment (davidot) [reply Userland/Libraries/LibJS/Parser.cpp]: Yeah that as well

File: Userland/Libraries/LibJS/Parser.h (hunk 2 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Parser.h]: TIL you can add friend functions as well - done :)

File: Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp]: Yep, missed that. Fixed.

File: Userland/Libraries/LibJS/Parser.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Parser.cpp]: ah, that's leftover from debugging. Initially I was passing parse_options, but turns out it doesn't consume the async token anymore if it already determined `function_kind = FunctionKind::Async` from those options :acid2:
Reverted :+1:

File: Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp (hunk 4 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/FunctionConstructor.cpp]: Nope, initially didn't compile with just auto but I think it was a missing include. Updated.

**Review by davidot — APPROVED** [review]
